File: lintcode/unique-binary-search-trees-ii.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/unique-binary-search-trees-ii/

"""
Definition of TreeNode:
"""
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None

class Solution:
    def work_with_recursion(self, istart, iend):
        ans = []
        if istart > iend:
            return ans

        for i in xrange(istart, iend+1):
            for j in self.work_with_recursion(istart, i-1) or [None]:
                for k in self.work_with_recursion(i+1, iend) or [None]:
                    root = TreeNode(i)
                    root.left = j
                    root.right = k
                    ans.append(root)

        return ans

    # @paramn n: An integer
    # @return: A list of root
    def generateTrees(self, n):
        # write your code here
        if n == 0:
            return [[]]

        return self.work_with_recursion(1, n)

if __name__ == '__main__':
    s = Solution()
    trees = s.generateTrees(3)

File: lintcode/unique-binary-search-trees-ii.cc
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @paramn n: An integer
     * @return: A list of root
     */
    vector<TreeNode *> generateTrees(int n) {
        // write your code here
        return this->dfs(1, n);
    }
private:
    vector<TreeNode *> dfs(int i, int j) {
        vector<TreeNode *> ans;
        if (i > j) {
            ans.push_back(NULL);
            return ans;
        }
        if (i == j) {
            ans.push_back(new TreeNode(i));
            return ans;
        }
        for (int p = i; p <= j; ++p) {
            vector<TreeNode *> ans_left = this->dfs(i, p-1);
            vector<TreeNode *> ans_right = this->dfs(p+1, j);
            for (int q = 0; q < ans_left.size(); ++q) {
                for (int r = 0; r < ans_right.size(); ++r) {
                    TreeNode *root = new TreeNode(p);
                    root->left = ans_left[q];
                    root->right = ans_right[r];
                    ans.push_back(root);
                }
            }
        }
        return ans;
    }
};
